Feed tools

JsseL opened the pull request:
1. Utilities for fetching Twitter and Facebook feeds.
2. Utilities to aggregate, order and display feeds.
3. Django CMS plugin for showing feeds on CMS site.
4. Management command for testing purposes.

JsseL committed:
Project test scaffolding

Basic things required to run tests.

JsseL committed:
Add funtionality to fetch social media feeds

Currently only Twitter and Facebook supported.

Utilities for fetching profile feeds and a uniformed API for accessing
feed items.

akx commented on cmsplugin_feed_ai/__init__.py:
```diff
@@ -0,0 +1,11 @@
+# ,----ooO--(_)-------.
+# | Please            |
+# |   don't feed the  |
+# |     TROLL's !     |
```
Get rid of the apostrophe and space before the exclamation mark. ;)

> TROLLS!

akx commented on cmsplugin_feed_ai/cms_plugins.py:
```diff
@@ -0,0 +1,39 @@
+    render_template = "cmsplugin_feed_ai/feed.html"
+
+    def get_feed(self, limit_per_service):
+        feed = []
```
Get rid of this . . .

akx commented on cmsplugin_feed_ai/cms_plugins.py:
```diff
@@ -0,0 +1,39 @@
+    def get_feed(self, limit_per_service):
+        feed = []
+        try:
+            feed = get_aggregated_profile_feed(
```
Make this `return get_...` . . .

akx commented on cmsplugin_feed_ai/cms_plugins.py:
```diff
@@ -0,0 +1,39 @@
+            pass
+        except GraphAPIError:
+            pass
+        return feed
```
And turn this into `return []`.

akx commented on cmsplugin_feed_ai/cms_plugins.py:
```diff
@@ -0,0 +1,39 @@
+    model = ProfileFeedPluginConf
+    render_template = "cmsplugin_feed_ai/feed.html"
+
+    def get_feed(self, limit_per_service):
```
Missing docstring.

akx commented on cmsplugin_feed_ai/cms_plugins.py:
```diff
@@ -0,0 +1,39 @@
+        feed = self.get_feed(plugin_conf.items_per_service)
+        context.update({
+            "feed": feed,
+            "placeholder": placeholder
```
Add a trailing comma, please.

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,171 @@
+
+    def __init__(self, feed_obj, profile_name, source):
+        """
+        Initializer display item from raw feed object provided by
```
> Initialize

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,171 @@
+        self.profile_name = profile_name
+        self.source = source
+        if source == "twitter":
+            self._init_from_twitter_object(feed_obj)
```
Mmm, these should be subclasses so one class won't need to know about different providers.

``` python
class FeedDisplayItem(object):
   ...

class TwitterFeedDisplayItem(FeedDisplayItem):
   source = "twitter"
   # ctor has logic for populating from twitter object

class FacebookFeedDisplayItem(...

```

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,171 @@
+    @property
+    def service_url(self):
+        """
+        URL for the source service
```
This can become a static property of the subclasses.

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,171 @@
+        return "%s/%s" % (self.service_url, self.profile_name)
+
+    @property
+    def permalink(self):
```
This can also be simply overridden in subclasses.

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,171 @@
+def get_feed_settings(key=None, default=None):
+    feed_settings = getattr(settings, "AI_FEED_SETTINGS", {})
+    if key not in feed_settings:
+        warn("Missing %r settings in AI_FEED_SETTINGS" % key)
```
> Missing %r setting

(not plural)

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,171 @@
+    return "cmsplugin_feed_ai_%s_feed" % service
+
+
+def _fetch_twitter_profile_feed(profile_name, limit):
```
Twitter-specific code should live in one submodule (`cmsplugin_feed_ai.providers.twitter`?) and Facebook-specific in another, I think.

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,171 @@
+    """
+    conf = get_feed_settings("twitter")
+    api = Twython(
+        conf.get("app_key", ""),
```
Use kwargs when passing these in to the Twython ctor, for clarity. Also, I guess you'll want it to fail with a KeyError instead of trying to pass in empty args, so `[]` instead of `.get`?

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,171 @@
+    :type service: Str
+    :ret type: List of FeedDisplayItems
+    """
+    key = get_feed_cache_key(service)
```
Since the data you're caching is variant on all 3 parameters (`service`, `profile_name` and `limit`), the cache key should definitely be too.

Right now, if I were to call `get_profile_feed('twitter', 'akx')`, then `get_profile_feed('twitter', 'POTUS')`, I'd still get my tweets, not the president's...

akx commented on cmsplugin_feed_ai/models.py:
```diff
@@ -0,0 +1,26 @@
+
+class ProfileFeedPluginConf(CMSPlugin):
+    """
+    Plugin model for lifting articles filtered by tags.
```
Ahem.

akx commented on setup.py:
```diff
@@ -16,6 +16,10 @@
     description='Social media feed plugin for Django CMS',
     install_requires=[
         'django-cms>=3.2,<3.4',
+        'django-soft-choice-fields>=0.3.1',
```
Not an actual dependency of the project.

akx commented on test_settings.py:
```diff
@@ -21,13 +37,9 @@
 
 LANGUAGES = [
-    ('en-us', 'English'),
+    ("en-us", "English"),
```
Don't mix quote styles within the file.

akx commented on test_settings.py:
```diff
@@ -1,6 +1,22 @@
+
+
 DEBUG = True
-SECRET_KEY = 'x'
```
Why did these need changing/removing?

JsseL replied:
I just had them already while you added another test settings into the repo. So pretty much just decided to use my existing version.

akx commented on tests/test_feed.py:
```diff
@@ -0,0 +1,52 @@
+from twython.exceptions import TwythonAuthError
+
+from cmsplugin_feed_ai.feed import FeedDisplayItem
+from test_data import facebook, twitter
```
This package seems to be missing. Also, I'd appreciate it if these were imported with less ambiguous local names.

``` python
from test_data import facebook as facebook_test_data
```

etc maybe?

akx commented on tests/test_plugins.py:
```diff
@@ -0,0 +1,25 @@
+
+
+@pytest.mark.django_db
+def test_article_list_plugin_html():
```
Ahem.

akx reviewed (changes requested):
First round of review done...

JsseL commented:
1st round of fixes according to comments.

akx commented on cmsplugin_feed_ai/__init__.py:
```diff
@@ -0,0 +1,11 @@
+#          (o o)
+# ,----ooO--(_)-------.
+# |                   |
+# |       Feed me     |
```
Better. ;)

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,190 @@
+        return "%s/posts/%s" % (self.profile_url, self.source_id.split("_")[-1])
+
+
+def get_feed_settings(key=None, default=None):
```
Can `key` really be `None`?

JsseL replied:
Removed unnecessary and unused default value.

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,190 @@
+
+
+def get_feed_cache_key(service, profile_name, limit):
+    """Return a unique cache key for given service"""
```
Docstring no longer up to date.

akx replied:
You use this function in exactly one place (and I foresee it won't be used elsewhere), so inline it.

JsseL replied:
Done

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,190 @@
+    if not feed:
+        feed = _fetch_profile_feed(service, profile_name, limit)
+        cache.set(key, feed, get_feed_settings("cache_time", DEFAULT_CACHE_TIME))
+    if service == "twitter":
```
This looks like it should be a dict elsewhere.

``` python
SERVICE_TO_DISPLAY_ITEM_CLASS_MAP = {
  'twitter': ...
}
```

That map could also be used for enumerating which services we support:

``` python
SUPPORTED_SERVICES = set(...map.keys())
```

akx replied:
(and said set can then be used where we now hard-code `"twitter"` and `"facebook"` :)

JsseL replied:
Done

akx commented on tests/test_feed.py:
```diff
@@ -0,0 +1,53 @@
+    item = TwitterFeedDisplayItem(feed_item, twitter_test_data.PROFILE_NAME)
+    assert item.source == "twitter"
+    assert isinstance(item.created_at, datetime)
+    assert item.service_url == "https://twitter.com"
```
I don't think checking the service url is much of an actually relevant test.

JsseL replied:
Not anymore at least. 😄

JsseL commented:
Refactored code, cleaned according to comments, fixed flake8 errors and ran isort.

akx commented on test_data/facebook.py:
```diff
@@ -0,0 +1,23 @@
+        'created_time': '2016-08-26T23:22:37+0000'
+    }],
+    'paging': {
+        'previous': 'https://graph.facebook.com/v2.7/424207551102424/feed?limit=2&since=1473786299&access_token=637781556382320|998ef1b90343a4d3b3ca7f0bf98e89c2&__paging_token=enc_AdDVKTHseqJB0tTHqyZC69aEz4brFyQ0iHQ781sAdYobWZBXiHdQJ2Swjm4JRUIxbqUFuJUZCDrSLpvlfIaY0t3KlgU8m6Ir2g8uZCTk57l3WKSDngZDZD&__previous=1',
```
Maybe remove all of these tokens (though I did check that these aren't valid URLs) from the text fixture data. Or replace them with `=x` or something anyway...

JsseL replied:
Old expired test tokens. But replaced now with dummy tokens.

akx commented on cmsplugin_feed_ai/cms_plugins.py:
```diff
@@ -0,0 +1,42 @@
+from cms.plugin_base import CMSPluginBase
+from cms.plugin_pool import plugin_pool
+from django.utils.translation import ugettext_lazy as _
+from facebook import GraphAPIError
```
Feels bad that the CMS plugin should have to know about the underlying API libraries.

Maybe add `.providers.excs`, and create a `FeedException` class that wraps these:

``` python
class FeedException(Exception):
    def __init__(self, exc):
        super().__init__('%s: %s' % (exc.__class__, exc))
        self.original_exception = exc
```

and then catch only those here? :)

JsseL replied:
Sounds good.

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,101 @@
+    return feed_settings.get(key, default)
+
+
+def _fetch_twitter_profile_feed(profile_name, limit):
```
This should live in `.providers.twitter` as well, and should likewise be mapped via a dict:

``` python
SERVICE_TO_FETCHER_MAP = {
    "facebook": ....,
    "twitter": fetch_twitter_profile_feed,
}
```

JsseL replied:
Yap.

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,101 @@
+    This method uses cache to prevent excess calls to the service APIs.
+    :param service: Service identifier
+    :type service: Str
+    :ret type: List of FeedDisplayItems
```
`:rtype: list[FeedDisplayItem]`

akx replied:
☝️

akx commented on cmsplugin_feed_ai/management/commands/test_social_feeds.py:
```diff
@@ -0,0 +1,39 @@
+        self.output_feeds(profile_name, limit)
+
+    def output_feeds(self, profile_name, limit):
+        twitter_feed = _fetch_profile_feed("twitter", profile_name, limit)
```
Once you've refactored the fetchers into a dict, this can just use that -- and then we're practically out of places where `twitter` and `facebook` are hard-coded! :)

akx commented on cmsplugin_feed_ai/models.py:
```diff
@@ -0,0 +1,26 @@
+        default=5,
+        verbose_name=_("number of items per service"),
+        help_text=_(
+            "Determines the max. number of items to be shown per social "
```
Come to think of it, I don't think this help text really provides any extra information c.f. the verbose name.

akx replied:
☝️

JsseL commented:
Refactored.

akx commented on cmsplugin_feed_ai/feed.py:
```diff
@@ -0,0 +1,60 @@
+def get_aggregated_profile_feed(services, limit_per_service=5):
+    """
+    Aggregate profile feeds from multiple services and
+    sort the feed according to creation times.
```
Missing `:rtype:` annotation.

akx commented on test_data/twitter.py:
```diff
@@ -0,0 +1,184 @@
+# -*- coding: utf-8 -*-
+# flake8: noqa
+"""
+Example data got from Twython.get_user_timeline mehtod.
```
method

akx commented on cmsplugin_feed_ai/cms_plugins.py:
```diff
@@ -0,0 +1,39 @@
+                limit_per_service=limit_per_service
+            )
+        except FeedException:
+            pass
```
You can just put the `return []` here.

akx reviewed (changes requested):
A couple more small nits, and then I'd also ask you to squash the changes as appropriate; intermediate changes followed by refactoring won't be required in the final history of the project.

akx commented:
Oh, and to fix the Travis fails, add `from __future__ import absolute_import` to `providers/facebook.py`.

JsseL committed:
Add CMS plugin for lifting profile feeds

Wire the previously implemented feed fetching utilities to a cms
plugin that can configure how many feed items will be fetched.

JsseL committed:
Add management command for testing feed fetching

Command to test everything works and is setup correctly. Also good for
debugging.

JsseL commented:
Ok, codecov seems to be only one disappointed now. Test coverage decrease isn't informative since earlier there wasn't any real tests.

akx commented:
Yeah, that's fine. Merging. Thanks for putting up with my complaints!

JsseL commented:
Less bug fixing later on 😉
